synapse/storage/appservice.py

```diff
@@ -174,6 +198,68 @@ def get_app_service_by_token(self, token, from_cache=True):
         # TODO: This should be JOINed with the application_services_regex table.
 
     @defer.inlineCallbacks
+    def get_app_service_rooms(self, service):
+        """Get a list of RoomsForUser for this application service.
+
+        Application services may be "interested" in lots of rooms depending on
+        the room ID, the room aliases, or the members in the room. This function
+        takes all of these into account and returns a list of RoomsForUser which
+        represent the entire list of room IDs that this application service
+        wants to know about.
+
+        Args:
+            service: The application service to get a room list for.
+        Returns:
+            A list of RoomsForUser.
+        """
+        # FIXME: This is assuming that this store has methods from
+        # RoomStore, DirectoryStore, RegistrationStore, RoomMemberStore which is
+        # a bad assumption to make as it makes testing trickier and coupling
+        # less obvious.
+
+        # get all rooms matching the room ID regex.
+        room_entries = yield self.get_all_rooms()  # RoomEntry list
+        matching_room_id_list = [
+            r.room_id for r in room_entries if
+            service.is_interested_in_room(r.room_id)
+        ]
+
+        # resolve room IDs for matching room alias regex.
+        room_alias_mappings = yield self.get_all_associations()
+        matching_alias_list = [
+            r.room_id for r in room_alias_mappings if
+            service.is_interested_in_alias(r.room_alias)
+        ]
+        room_ids_matching_alias_or_id = set(
+            matching_room_id_list + matching_alias_list
+        )
+
+        # get all rooms for every user for this AS. This is scoped to users on
+        # this HS only.
+        user_list = yield self.get_all_users()
+        user_list = [
+            u["name"] for u in user_list if
+            service.is_interested_in_user(u["name"])
+        ]
+        rooms_for_user_matching_user_id = []  # RoomsForUser list
+        for user_id in user_list:
+            rooms_for_user = yield self.get_rooms_for_user(user_id)
+            rooms_for_user_matching_user_id += rooms_for_user
+        rooms_for_user_matching_user_id = set(rooms_for_user_matching_user_id)
+
+        # make RoomsForUser tuples for room ids and aliases which are not in the
+        # main rooms_for_user_list - e.g. they are rooms which do not have AS
+        # registered users in it.
+        known_room_ids = [r.room_id for r in rooms_for_user_matching_user_id]
+        missing_rooms_for_user = [
+            RoomsForUser(r, service.sender, "join") for r in
+            room_ids_matching_alias_or_id if r not in known_room_ids
+        ]
+        rooms_for_user_matching_user_id |= set(missing_rooms_for_user)
+
+        defer.returnValue(rooms_for_user_matching_user_id)
```

Comment:
Should this all be done in a single transaction rather than jumping backwards and forwards between threads? It would also allow us to more easily assess what the performance is like using the transaction performance logging.